mkosi/qemu.py

```diff
@@ -702,14 +702,16 @@ def finalize_drive(drive: QemuDrive) -> Iterator[Path]:
 
 @contextlib.contextmanager
 def finalize_state(config: Config, cid: int) -> Iterator[None]:
-    (INVOKING_USER.runtime_dir() / "machine").mkdir(parents=True, exist_ok=True)
+    mkosi_pid = os.getpid()
+    machine_folder = INVOKING_USER.runtime_dir() / "machine"
+    machine_folder.mkdir(parents=True, exist_ok=True)
 
     if INVOKING_USER.is_regular_user():
         os.chown(INVOKING_USER.runtime_dir(), INVOKING_USER.uid, INVOKING_USER.gid)
-        os.chown(INVOKING_USER.runtime_dir() / "machine", INVOKING_USER.uid, INVOKING_USER.gid)
+        os.chown(machine_folder, INVOKING_USER.uid, INVOKING_USER.gid)
 
-    with flock(INVOKING_USER.runtime_dir() / "machine"):
-        if (p := INVOKING_USER.runtime_dir() / "machine" / f"{config.machine_or_name()}.json").exists():
+    with flock(machine_folder):
+        if (p := machine_folder / f"{config.machine_or_name()}-{mkosi_pid:d}.json").exists():
```

Comment:
The `:d` is not necessary, numbers are by default formatted as decimal numbers. 
```suggestion
        if (p := machine_folder / f"{config.machine_or_name()}-{mkosi_pid}.json").exists():
```

Reply:
Locking is done on the machine folder (typically `/run/user/$uid/mkosi/machine`). This PR does not change that.

Is there any reason no two `mkosi` instances should not use the same machine name?
If the only reason is that it's later used by `mkosi` to ssh into the VM, then the current approach is fine.

---

> What kind of crashes sidestep the `finally`?

I'm not sure if it's just `ctrl+C` at the wrong time or if it's just when I try booting an image that crashes or if I quit the terminal without doing `ctrl+C` but I can't seem to just be able to `mkosi qemu` for a full day without having to manually clean the machine state directory. It's genuinely happening a lot.

Reply:
@gdonval Are you on the latest commit on main? We recently added logic to handle `SIGHUP`, that could maybe fix your issue.

Reply:
No, I'm on 24.3.

At any rate, I'm a bit confused... Is the `if ...: die()` here a simple way to prevent spurious overwrites/deletions or is there any value in making sure machine names are unique across the machine?

If it's the first one, I think my approach is valid. If it's the second one, maybe let's just not use hostname as **default** machine name value?

Reply:
> At any rate, I'm a bit confused... Is the if ...: die() here a simple way to prevent spurious overwrites/deletions or is there any value in making sure machine names are unique across the machine?

The `SIGHUP` handling we have on git main will likely reduce the number of times you run into this as the files in /run/ will be properly cleaned up when `mkosi qemu` crashes.

> If it's the first one, I think my approach is valid. If it's the second one, maybe let's just not use hostname as default machine name value?

@gdonval The machine name is used to identify the target for `mkosi ssh` so making it random would make that harder. We can probably pick a better name than `default` though and allow disabling "registering" the machine.

Reply:
👍 for `SIGHUP`.

@DaanDeMeyer, so

1. The machine name is not a lock, therefore its existence should not make new instances die.
2. That's a `mjosi ssh` concern, therefore it should not affect new `mkosi qemu` calls.

I suggest:

1. either don't write a state file, unless `--machine` is specified;
2. or write machine named with a counter appended (similar to this PR) so that `mkosi qemu` doesn't crash.

In the latter case, `qemu ssh` can look for a precise machine name (with counter) or could look at the highest counter for sma specific machine name stem.

Reply:
@gdonval I prefer `mkosi ssh` to work out of the box without having to specify `--machine`. What about using the working directory name as the machine name instead of `default`? That should drastically reduce the number of conflicts.

Reply:
@DaanDeMeyer, you prefer 

> `mkosi ssh` to work without having to specify machine

and suggest using the folder name. 

Am I right to assume you would call `mkosi ssh` only from the current `mkosi` folder then? 

In that case, why don't we write a softlink reference somewhere appropriate in the current `mkosi` project folder? State would still be written centrall (albeit with a convenient distinguisher) and `mkosi ssh` would follow the local link to that folder's default session.